lib/ansible/modules/cloud/misc/qemu_img.py

```diff
@@ -0,0 +1,217 @@
+    description:
+    - Whether the image is allowed grow
+  shrink:
+    choices: [ "yes", "no" ]
+    defaults: "no"
+    description:
+    - Whether the image is allowed shrink
+  state:
+    choices: [ "absent", "present"]
+    description:
+    - If the image should be present or absent - default present
+notes:
+  - This module does not change the type/format of the image
+  - This module does not take snapshots, and should be implemented seperate
+'''
+
+EXAMPLES = '''
+# Create a raw image of 5M.
+- qemu_img:
+    dest: /tmp/testimg
+    size: 5M
+    format: raw
+
+# Enlarge the image to 6G.
+- qemu_img:
+    dest: /tmp/testimg
+    size: 6G
+    format: qcow2
+
+# Shrink the image by 3G
+    code: qemu_img
+    dest: /tmp/testing
+    size: -3G
+    shrink: yes
+    format: qcow2
+
+# Remove the image
+- qemu_img:
+    dest: /tmp/testimg
+    state: absent
+'''
+
+RETURN = '''
+# create qemu image
+present:
+  type: string
+  sample: "success"
+  returned: success
+# remove qemu image
+absent:
+  type: string
+  sample: "success"
+  returned: success
+# resize qemu image
+resize:
+  type: string
+  sample: "success"
+  returned: success
+'''
+
+import os
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.pycompat24 import get_exception
+
+def main():
+
+    module = AnsibleModule(
+        argument_spec = dict(
+            dest = dict(type='str', required=True),
+            options = dict(type='str', default='preallocation=metadata'),
+            format = dict(type='str', default='qcow2'),
+            size = dict(type='str'),
+            grow= dict(type="bool", default=True),
+            shrink = dict(type="bool", default=False),
+            state = dict(type='str', choices=['absent', 'present'], default='present'),
+        ),
+        supports_check_mode = True,
+        required_if=[
+            ( "state", "present", [ "size" ] ),
+        ],
+    )
+
+    result = dict(
+        changed = False,
+    )
+
+    state = module.params['state']
+    dest = module.params['dest']
+    img_format = module.params['format']
+    img_options = module.params['options']
+    size = module.params['size']
+    grow = module.params['grow']
+    shrink = module.params['shrink']
+
+    qemu_img = module.get_bin_path('qemu-img', True)
+
+    if state == 'present':
+        if not os.path.exists(dest):
+            if not module.check_mode:
+                if not img_options:
+                    rc, stdout, stderr = module.run_command('%s create -f %s "%s" %s' % (qemu_img, img_format, dest, size), check_rc=True)
+                else:
+                    rc, stdout, stderr = module.run_command('%s create -f %s -o "%s" "%s" %s' % (qemu_img, img_format, img_options, dest, size), check_rc=True)
+            result['changed'] = True
+        else:
+            try:
+                size_unit = str(size[-1:])
+                size = size[:-1]
+            except:
+                size_unit = 'b'
```

Comment:
I don't think this is working correctly, did you test it for all possible inputs ?

- First of all the size_unit is only correct if it is a known unit.
- And if that's not true, and it's not a digit, you have to escalate.
- If it is a digit, you likely won't have a unit and then you could assume bytes.